server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -743,25 +749,20 @@ protected void doRun() throws Exception {
                                     sendFileExecutor.execute(this); // fork off from the network thread
                                 }
                             },
-                            // need to fork as `handleErrorOnSendFiles` might read some files which should not happen on the network thread
-                            e -> sendFileExecutor.execute(new ActionRunnable<>(this.listener) {
-                                @Override
-                                protected void doRun() throws Exception {
-                                    cancellableThreads.execute(semaphore::acquire);
-                                    try (Releasable ignored = semaphore::release) {
-                                        handleErrorOnSendFiles(store, e, new StoreFileMetaData[]{chunk.md});
-                                        throw e;
-                                    }
+                            e -> {
+                                if (error.compareAndSet(null, Tuple.tuple(chunk.md, e)) && semaphore.tryAcquire()) {
```

Comment:
It seems we're just dropping all but the first error here, maybe we should at least log subsequent errors?

Reply:
addressed in 6f8f7a4.